src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -112,9 +112,35 @@ public int Count
         
         public override int GetHashCode()
         {
-            return null == _array
-                        ? 0
-                        : _array.GetHashCode() ^ _offset ^ _count;
+            if (_array == null)
+            {
+                return 0;
+            }
+            
+#if !FEATURE_CORECLR
+            // Old implementation: Xor everything.
+            // This isn't optimal since it treats _offset and _count interchangeably,
+            // but we keep it around in the full framework for compat.
+            return _array.GetHashCode() ^ _offset ^ _count;
+#else
+            // Modified Bernstein hash: start with 5381,
+            // multiply by 33, xor with the next hash.
+            // NOTE: (hash << 5) + hash is equal to hash * 33.
+            
+            int hash = (33 * 5381) ^ _offset;
+            hash = ((hash << 5) + hash) ^ _count;
+            
+            int arrayHash = _array.GetHashCode();
+            // Note quite sure why, but it looks like the result of object.GetHashCode
+            // only returns 26 bits in the return value- the upper 6 are set to zero.
+            // See src/vm/syncblk.h for more.
+            // As a result, we take the array hash and shift it 6 to the left
```

Comment:
@VSadov It is [the same](https://github.com/dotnet/corert/blob/4376c55dd018d8b7b9fed82449728711231ba266/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L53) on .NET Native though, although yeah I agree that it does seem like somewhat of an implementation detail. If it's alright with you, I think we should change the approach to switch the lower 8 bits before XORing with the array, like this:

```cs
int hash = (33 * 5381) ^ _offset;
hash = ((hash << 5) + hash) ^ _count;
hash = (hash << 24) | (hash >> 8); // since _offset and _count are typically low,
// this will move their bits to the top of the integer
hash ^= _array.GetHashCode();
```

This way we don't waste the upper few bits due to `object.GetHashCode`, and still stay resilient in the future if it's changed to return more than 26 bits.